### PR TITLE
Bug 1293066 – firefox: URL scheme bypasses localhost navigation blocking

### PR DESCRIPTION
**Pull request by varkor**
Simply block any local requests opened via the firefox:// URI scheme.

### CODE REVIEW

**Comment by thebnich** [comment]
Hm, we should already be blocking unprivileged requests, and loading arbitrary URLs from the `firefox://` scheme should definitely *not* privileged, so I think the real fix for this problem is a bit deeper. 

This load eventually ends up calling `openURLInNewTab` in `BVC`, which is where we wrap the URL in a `PrivilegedRequest`. It looks like I did this as part of [bug 1268871](https://bugzilla.mozilla.org/show_bug.cgi?id=1268871), but I was a bit careless since I didn't thoroughly look at all the ways we can end up routing through that method.

I think the best way to fix this would to be explicit about exactly when we need a privileged request, denying it in all other cases. I'd probably either 1) change `openURLInNewTab` to take an `NSURLRequest` instead of an `NSURL`, or 2) add an `isPrivileged` parameter to the existing method.

**Comment by varkor** [comment]
Yeah, that makes sense. I've added an `isPrivileged` argument. Now, if an invalid (e.g. reader mode) URL is attempted to be opened from the `firefox://` URL scheme, it will simply open a blank tab. Is this acceptable, or do we want to warn the user that the URL they tried to load was faulty in some way?

**Comment by thebnich** [comment]
That should be fine for now...I don't expect many apps will be trying to share these schemes.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -577,7 +577,7 @@ class BrowserViewController: UIViewController {
 
         if shouldShowWhatsNewTab() {
             if let whatsNewURL = SupportUtils.URLForTopic("new-ios") {
-                self.openURLInNewTab(whatsNewURL)
+                self.openURLInNewTab(whatsNewURL, isPrivileged: true)
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: This should be `false` (the support URL is on https://support.mozilla.org).

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 2]
```diff
@@ -1068,7 +1068,7 @@ class BrowserViewController: UIViewController {
 
     func openBlankNewTabAndFocus(isPrivate isPrivate: Bool = false) {
         popToBVC()
-        openURLInNewTab(nil, isPrivate: isPrivate)
+        openURLInNewTab(nil, isPrivate: isPrivate, isPrivileged: true)
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Should be `false` since this is just a blank tab.

Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Edit: Nevermind; I think `nil` falls back to the user's homepage choice, which could be local.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 3]
```diff
@@ -1411,7 +1411,7 @@ extension BrowserViewController: MenuActionDelegate {
 
 extension BrowserViewController: SettingsDelegate {
     func settingsOpenURLInNewTab(url: NSURL) {
-        self.openURLInNewTab(url)
+        self.openURLInNewTab(url, isPrivileged: true)
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Can make this `false`; I don't think settings opens any local pages.

**Comment by thebnich** [comment]
Thanks for the updated PR!

**Pull request merged**